arborist/resource.go

```diff
@@ -0,0 +1,182 @@
+	// a directory, the subresources are the immediate contents of this
+	// directory.
+	subresources map[*Resource]struct{}
+}
+
+func validateResourceName(name string) error {
+	if !utf8.Valid([]byte(name)) {
+		return nameError(name, "resource", "only UTF8 allowed")
+	}
+
+	if strings.Contains(name, "/") {
+		return nameError(name, "resource", "can't use reserved character '/'")
+	}
+
+	return nil
+}
+
+// NewResource sets up a new resource. If the parent is given then this
+// resource is attached as a subresource, and its path will be computed
+// accordingly (appending the name for this resource as the new last segment
+// after the path of the parent node). Returns an error if the name is invalid
+// as according to `validateResourceName`.
+//
+// NOTE: a resource may point to a parent resource, but until the engine runes
+// `addResource` (or the parent is otherwise modified), the parent resource
+// will not have a pointer to this resource as a subresource.
+func NewResource(
+	name string,
+	description string,
+	parent *Resource,
+	subresources map[*Resource]struct{},
+) (*Resource, error) {
+	if err := validateResourceName(name); err != nil {
+		return nil, err
+	}
+
+	var pathSegments []string
+	if parent != nil {
+		// For this case we have to copy the values out of the parent path
+		// into this one.
+		pathSegments = make([]string, len(parent.pathSegments)+1)
+		for i, p := range parent.pathSegments {
+			pathSegments[i] = p
+		}
+		pathSegments[len(parent.pathSegments)] = name
+	} else {
+		pathSegments = []string{name}
+	}
+
+	path := pathString(pathSegments)
+
+	if subresources == nil {
+		subresources = make(map[*Resource]struct{})
+	}
+
+	resource := &Resource{
+		name:         name,
+		path:         path,
+		pathSegments: pathSegments,
+		description:  description,
+		parent:       parent,
+		subresources: subresources,
+	}
+
+	if parent != nil {
+		parent.addSubresource(resource)
+	}
+
+	return resource, nil
+}
+
+func (resource *Resource) addSubresource(sub *Resource) {
+	resource.subresources[sub] = struct{}{}
+	sub.parent = resource
+}
+
+func (resource *Resource) rmSubresource(sub *Resource) {
+	delete(resource.subresources, sub)
+	sub.parent = nil
+}
+
+func (resource *Resource) equals(other *Resource) bool {
+	pathLen := len(resource.path)
+	otherPathLen := len(other.path)
+	if pathLen != otherPathLen {
+		return false
+	}
+	for i := 0; i < pathLen; i++ {
+		if resource.path[i] != other.path[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// Traverse does a basic BFS starting at the given resource and traversing
+// through all the subresources starting from that resource, writing the output
+// to a channel. It receives a channel (`done`) which can indicate to cut off
+// the output and return early.
+//
+// NOTE that the traversal order is not guaranteed to be anything in particular.
+func (resource *Resource) traverse(done chan struct{}) <-chan *Resource {
```

Comment:
Ooh, channels. Neat!

Reply:
Yeah, this I think was a cool place to use them and easy enough to add :smile: other places for now I'm considering as premature optimization...